File: core/src/main/java/glitch/api/ws/events/Event.java
package glitch.api.ws.events;

import glitch.api.AbstractWebSocketService;
import lombok.AccessLevel;
import lombok.Data;
import lombok.RequiredArgsConstructor;
import lombok.ToString;

import java.time.Instant;
import java.util.UUID;

/**
 * All events must be extends this class if they want to be emitted via {@link reactor.core.publisher.FluxProcessor}
 * @param <S> client extended {@link AbstractWebSocketService}
 */
@Data
@RequiredArgsConstructor(access = AccessLevel.PROTECTED)
public abstract class Event<S extends AbstractWebSocketService<S>> implements IEvent<S> {
    @ToString.Exclude
    private final S client;
    private final Instant createdAt;
    private final UUID eventId;

    protected Event(S client) {
        this(client, Instant.now());
    }

    protected Event(S client, Instant createdAt) {
        this(client, createdAt, UUID.randomUUID());
    }
}

File: core/src/main/kotlin/io/glitchlib/model/event.kt
package io.glitchlib.model

import io.glitchlib.GlitchClient
import java.util.Date
import java.util.UUID

interface IEvent : GlitchObject, CreatedAt {
    val eventId: UUID
        get() = UUID.randomUUID()
    override val createdAt: Date
        get() = Date()
}

data class OpenEvent(
    override val client: GlitchClient
) : IEvent

data class CloseEvent(
    val status: Status,
    override val client: GlitchClient
) : IEvent {
    data class Status(
        val code: Int,
        val reason: String?
    )
}

data class PingEvent(
    override val client: GlitchClient
) : IEvent

data class PongEvent(
    override val client: GlitchClient
) : IEvent